airbyte-integrations/connectors/source-zuora/source_zuora/zuora_client.py

```diff
@@ -0,0 +1,270 @@
+
+        if q_type not in valid_types:
+            raise ZOQLQueryNotValid(q_type, valid_types)
+        elif q_type == "select":
+            end_date = pendulum.now().to_datetime_string() if end_date is None else end_date
+            base_query[
+                "query"
+            ] = f"select * from {obj} where {date_field} >= TIMESTAMP '{start_date}' and {date_field} <= TIMESTAMP '{end_date}'"
+        elif q_type == "describe":
+            base_query["query"] = f"DESCRIBE {obj}"
+        elif q_type == "show_tables":
+            base_query["query"] = "SHOW TABLES"
+        return base_query
+
+    # SUBMIT: data_query_job using data_query
+    def _submit_dq_job(self, dq_query: Dict) -> str:
+        submit_job = self._make_request(method="POST", url=f"{self.url_base}/query/jobs", payload=dq_query)
+        # self.logger.debug(submit_job.json()["data"]["id"])
+        return submit_job.json()["data"]["id"]
+
+    # CHECK: the submited data_query_job status
+    def _check_dq_job_status(self, dq_job_id: str) -> Dict:
+        # Define the job error statuses
+        errors = ["failed", "canceled", "aborted"]
+        # Error msg: the cursor_field cannot be resolved
+        cursor_error = "Column 'updateddate' cannot be resolved"
+        # Error msg: cannot process object
+        obj_read_error = "failed to process object"
+
+        status = None
+        success = "completed"
+        while status != success:
+            response = self._make_request(url=f"{self.url_base}/query/jobs/{dq_job_id}")
+            dq_job_check = response.json()
+            status = dq_job_check["data"]["queryStatus"]
+            if status in errors and cursor_error in dq_job_check["data"]["errorMessage"]:
+                raise ZOQLQueryFieldCannotResolve
+            elif status in errors and obj_read_error in dq_job_check["data"]["errorMessage"]:
+                raise ZOQLQueryCannotProcessObject
+            elif status in errors:
+                raise ZOQLQueryFailed(response)
+        return dq_job_check
+
+    # GET: data_query_job result
+    def _get_data_from_dq_job(self, response: requests.Response) -> List:
+        dq_job_data = requests.get(response["data"]["dataFile"])
+        dq_job_data.raise_for_status()
+        data = [j.loads(line) for line in dq_job_data.text.splitlines()]
+        return data or []
+
+    # Warpper function for `Query > Submit > Check > Get`
+    def _get_data(
+        self, q_type: str = "select", obj: str = None, date_field: str = None, start_date: str = None, end_date: str = None
+    ) -> Iterable:
+        query = self._make_dq_query(q_type, obj, date_field, start_date, end_date)
+        # self.logger.debug(f"{query}")
+        submit = self._submit_dq_job(query)
+        check = self._check_dq_job_status(submit)
+        get = self._get_data_from_dq_job(check)
+        yield from get
+
+    # Warper for _get_data using date-slices as pages
+    def _get_data_with_date_slice(
```

Comment:
create a list of slices and then in 1 slice run  '_get_data'

Reply:
I prefer to dynamically build the slices because if the sync goes a long time, we need to have the data up to `NOW`, basically, if we specify the slices using a strict list of slices, the last date-slice would have `end date < NOW`, and we still need to make another fetch to refresh the data will exactly now. Probably, I'm wrong, but building them up to date. is better here,  since there are a lot of streams and sometimes the incremental refresh would take up to 5-10 min.